### PR TITLE
Avoid interior mutability in `TyLoweringContext`

### PR DESCRIPTION
**Pull request by ChayimFriedman2**
This requires some serious code juggling.

Honestly I'm not sure this is worth it, it's just incredibly hard to fit inherited mutability into types that were not designed into it. This does have some nice advantages but some code becomes strictly worse. I'll leave that to the reviewer's opinion.

### CODE REVIEW

File: crates/hir-ty/src/lower.rs [hunk 1]
```diff
@@ -72,47 +72,32 @@ use crate::{
...
-    /// we're grouping the mutable data (the counter and this field) together
-    /// with the immutable context (the references to the DB and resolver).
-    /// Splitting this up would be a possible fix.
-    Opaque(RefCell<Arena<ImplTrait>>),
-    Param(Cell<u16>),
-    Variable(Cell<u16>),
-    Disallowed,
+    /// complicated).
+    mode: ImplTraitLoweringMode,
+    // This is structured as a struct with fields and not as an enum because it helps with the borrow checker.
```
Comment (Veykril) [inline crates/hir-ty/src/lower.rs:81]: I take it this is the part you are not too fond of here?

File: crates/hir-ty/src/lower.rs [hunk 2]
```diff
@@ -746,6 +717,8 @@ impl<'a> TyLoweringContext<'a> {
             res,
             Some(segment.name.clone()),
             move |name, t, associated_ty| {
+                let generics = self.generics().unwrap();
```
Comment (Veykril) [inline crates/hir-ty/src/lower.rs:720]: This seems unnecessary, we already have the generics in scope from line 740/711 above (or is that due to borrowing issues?)

Comment (Veykril) [reply crates/hir-ty/src/lower.rs:720]: Ah nvm it likely is

File: crates/hir-ty/src/lower.rs [hunk 3]
```diff
@@ -1049,11 +1026,11 @@ impl<'a> TyLoweringContext<'a> {
     }
 
     pub(crate) fn lower_where_predicate<'b>(
-        &'b self,
+        &'b mut self,
         where_predicate: &'b WherePredicate,
         &def: &GenericDefId,
         ignore_bindings: bool,
-    ) -> impl Iterator<Item = QuantifiedWhereClause> + 'b {
+    ) -> impl Iterator<Item = QuantifiedWhereClause> + use<'a, 'b> {
```
Comment (Veykril) [inline crates/hir-ty/src/lower.rs:1033]: new `use` feature already coming in very useful 🎉

File: crates/hir-ty/src/lower.rs [hunk 4]
```diff
@@ -1376,16 +1352,16 @@ impl<'a> TyLoweringContext<'a> {
...
         let self_ty = TyKind::BoundVar(BoundVar::new(DebruijnIndex::INNERMOST, 0)).intern(Interner);
         let predicates = self.with_shifted_in(DebruijnIndex::ONE, |ctx| {
-            let mut predicates: Vec<_> = bounds
-                .iter()
-                .flat_map(|b| ctx.lower_type_bound(b, self_ty.clone(), false))
-                .collect();
+            let mut predicates = Vec::new();
+            for b in bounds {
+                predicates.extend(ctx.lower_type_bound(b, self_ty.clone(), false));
+            }
```
Comment (Veykril) [inline crates/hir-ty/src/lower.rs:1362]: I take it there iter -> for loop desugarings are another annoyance you refer to, I don't immediately see why that's needed mind explaining that?

**Review by Veykril — APPROVED** [review]
In my opinion this is still an improvement even if it worsens some other parts of the codebase. Interior mutability is generally more difficult to reason about in my eyes, so a +1 from me

File: crates/hir-ty/src/lower.rs (hunk 1 above)
Comment (ChayimFriedman2) [reply crates/hir-ty/src/lower.rs:81]: This is actually quite okay, the part I don't like is having to convert iterator chains to loops.

File: crates/hir-ty/src/lower.rs (hunk 4 above)
Comment (ChayimFriedman2) [reply crates/hir-ty/src/lower.rs:1362]: Without it you get "captured variables cannot escape `FnMut` closure".

Comment (Veykril) [reply crates/hir-ty/src/lower.rs:1362]: Ah right, it returns an iterator borrowing from the environment ugh, that's unfortunate

Comment (ChayimFriedman2) [reply crates/hir-ty/src/lower.rs:1362]: Yeah you need lending fn to solve that.

**Pull request merged**